### PR TITLE
KAFKA-9373: Improve shutdown performance via lazy accessing the offset and time indices

### PR DESCRIPTION
**Pull request by efeg**
KAFKA-7283 enabled lazy mmap on index files by initializing indices on-demand rather than performing costly disk/memory operations when creating all indices on broker startup. This helped reducing the startup time of brokers. However, segment indices are still created on closing segments, regardless of whether they need to be closed or not.

This patch:
* Improves shutdown performance via lazy accessing the offset and time indices.
* Eliminates redundant disk accesses and memory mapped operations while deleting or renaming files that back segment indices.
* Prevents illegal accesses to underlying indices of a closed segment, which would lead to memory leaks due to recreation of the underlying memory mapped objects.

In our evaluations in a cluster with 31 brokers, where each broker has 13K to 20K segments, we observed up to 2 orders of magnitude faster LogManager shutdown times with this patch -- i.e. dropping the LogManager shutdown time of each broker from 10s of seconds to 100s of milliseconds.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/log/LazyIndex.scala [hunk 1]
```diff
@@ -17,27 +17,32 @@
...
   * A wrapper over an `AbstractIndex` instance that provides a mechanism to defer loading (i.e. memory mapping) the
   * underlying index until it is accessed for the first time via the `get` method.
   *
-  * This is an important optimization with regards to broker start-up time if it has a large number of segments.
+  * This is an important optimization with regards to broker start-up and shutdown time if it has a large number of segments.
+  * It prevents illegal accesses to the underlying index after closing the index, which might otherwise lead to memory
+  * leaks due to recreation of underlying memory mapped object.
   *
-  * Methods of this class are thread safe. Make sure to check `AbstractIndex` subclasses documentation
-  * to establish their thread safety.
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/LazyIndex.scala:34]: Why has the thread safety comment been removed?

**Comment by ijuma** [comment]
Nice optimization. I think we should consider delegating to the underlying implementations instead of adding logic to the wrapper.

**Comment by ijuma** [comment]
@efeg Any thoughts on the above?

**Comment by efeg** [comment]
@ijuma Thanks for the review and sorry for the delay on addressing the comments! -- I will try to update the PR by next week.

**Comment by ijuma** [comment]
@efeg thanks, looking forward to it.

**Comment by ijuma** [comment]
@efeg Seems like you have been busy. Since this is a very useful change, I went ahead and submitted a new pull request with the approach I suggested (with you as co-author):

https://github.com/apache/kafka/pull/8346

**Comment by ijuma** [comment]
Thanks for the PR, we merged it via https://github.com/apache/kafka/pull/8346.

**Comment by efeg** [comment]
Thanks for the updated PR @ijuma!

**Pull request closed without merging**